### PR TITLE
[SPARK-38796][SQL] Update to_number and try_to_number functions to restrict S and MI sequence to start or end only

### PR DESCRIPTION
**Pull request by dtenedor**
### What changes were proposed in this pull request?

Update `to_number` and `try_to_number` functions to restrict MI sequence to start or end only.

This satisfies the following specification:

```
to_number(expr, fmt)
fmt
  { ' [ MI | S ] [ L | $ ]
      [ 0 | 9 | G | , ] [...]
      [ . | D ]
      [ 0 | 9 ] [...]
      [ L | $ ] [ PR | MI | S ] ' }
```

### Why are the changes needed?

After reviewing the specification, this behavior makes the most sense.

### Does this PR introduce _any_ user-facing change?

Yes, a slight change in the behavior of the format string.

### How was this patch tested?

Existing and updated unit test coverage.

### CODE REVIEW

**Comment by cloud-fan** [comment]
LGTM, can we update the function doc as well?

**Comment by dtenedor** [comment]
> LGTM, can we update the function doc as well?

@cloud-fan this is done. It looks like the `On pull requests / Build and test (pull_request_target)` check is failing due to the same bug in the Spark Github REST API that is affecting many other PRs. Once that is fixed, the test should pass.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/ToNumberParser.scala [hunk 1]
```diff
@@ -49,33 +49,58 @@ object ToNumberParser {
   final val WRAPPING_ANGLE_BRACKETS_TO_NEGATIVE_NUMBER_END = 'R'
 
   // This class represents one or more characters that we expect to be present in the input string
-  // based on the format string.
-  abstract class InputToken()
+  // based on the format string. The toString method returns a representation of each token suitable
+  // for use in error messages.
+  abstract class InputToken() {
+    def toString: String
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/ToNumberParser.scala:55]: we don't need this, `toString` is built-in API for all java classes.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/numberFormatExpressions.scala [hunk 2]
```diff
@@ -46,8 +46,10 @@ import org.apache.spark.unsafe.types.UTF8String
            grouping separator relevant for the size of the number.
          '$': Specifies the location of the $ currency sign. This character may only be specified
            once.
-         'S': Specifies the position of a '-' or '+' sign (optional, only allowed once).
-         'MI': Specifies that 'expr' has an optional '-' sign, but no '+' (only allowed once).
+         'S': Specifies the position of a '-' or '+' sign (optional, only allowed once at the
+           beginning or end of the format string).
+         'MI': Specifies that 'expr' has an optional '-' sign, but no '+' (only allowed once at the
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/numberFormatExpressions.scala:51]: Shall we make the doc of `S` and `MI` more consistent? The only difference is `S` allows `+` but `MI` does not.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/ToNumberParser.scala (hunk 1 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/ToNumberParser.scala:55]: that said, we can override `toString` in subclasses already.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/ToNumberParser.scala [hunk 3]
```diff
@@ -49,33 +49,58 @@ object ToNumberParser {
...
+  }
   // Represents one closing angle bracket (>).
-  case class ClosingAngleBracket() extends InputToken
+  case class ClosingAngleBracket() extends InputToken {
+    override def toString: String = "PR"
+  }
   // Represents any unrecognized character other than the above.
-  case class InvalidUnrecognizedCharacter(char: Char) extends InputToken
+  case class InvalidUnrecognizedCharacter(char: Char) extends InputToken {
+    override def toString: String = s"character $char"
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/ToNumberParser.scala:102]:
```suggestion
    override def toString: String = s"character '$char'"
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/ToNumberParser.scala [hunk 4]
```diff
@@ -340,21 +322,65 @@ class ToNumberParser(numberFormat: String, errorOnFail: Boolean) extends Seriali
...
+      }
+      if (formatTokenIndex < formatTokens.length &&
+        allowed.exists(tokensMatch(_, formatTokens(formatTokenIndex)))) {
+        formatTokenIndex += 1
+      }
     }
+    if (formatTokenIndex < formatTokens.length) {
+      return s"Unexpected ${formatTokens(formatTokenIndex).toString} found in the format string " +
+        s"'$numberFormat'; the structure of the format string must match: " +
+        "[MI] [S] [$] [0|9|G|,]* [.|D] [0|9]* [$] [PR|MI|S]"
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/ToNumberParser.scala:380]: Do we require the ordering of `MI`, `S` and `$`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/ToNumberParser.scala (hunk 1 above)
Comment (dtenedor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/ToNumberParser.scala:55]: Good point, done.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/numberFormatExpressions.scala (hunk 2 above)
Comment (dtenedor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/numberFormatExpressions.scala:51]: Sure, SG, done.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/ToNumberParser.scala (hunk 3 above)
Comment (dtenedor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/ToNumberParser.scala:102]: Done.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/ToNumberParser.scala (hunk 4 above)
Comment (dtenedor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/ToNumberParser.scala:380]: Yes, the format check now enforces this ordering. I updated it to remove the ordering constraint for `MI` and `S` at the start of the format string per offline conversation.

**Comment by cloud-fan** [comment]
The test build: https://github.com/dtenedor/spark/actions/runs/2185310585

**Comment by cloud-fan** [comment]
thanks, merging to master/3.3!

**Pull request closed without merging**